module/zfs/dsl_bookmark.c

```diff
@@ -175,25 +299,20 @@ dsl_bookmark_create_check(void *arg, dmu_tx_t *tx)
 
 	for (nvpair_t *pair = nvlist_next_nvpair(dbca->dbca_bmarks, NULL);
 	    pair != NULL; pair = nvlist_next_nvpair(dbca->dbca_bmarks, pair)) {
-		dsl_dataset_t *snapds;
-		int error;
 
-		/* note: validity of nvlist checked by ioctl layer */
-		error = dsl_dataset_hold(dp, fnvpair_value_string(pair),
-		    FTAG, &snapds);
-		if (error == 0) {
-			error = dsl_bookmark_create_check_impl(snapds,
-			    nvpair_name(pair), tx);
-			dsl_dataset_rele(snapds, FTAG);
-		}
+		char *new = nvpair_name(pair);
+		char *source = fnvpair_value_string(pair);
+		int error = dsl_bookmark_create_check_impl(dp, new, source);
+
 		if (error != 0) {
-			fnvlist_add_int32(dbca->dbca_errors,
-			    nvpair_name(pair), error);
 			rv = error;
+			if (dbca->dbca_errors != NULL)
+				fnvlist_add_int32(dbca->dbca_errors,
+				    new, error);
 		}
 	}
 
-	return (rv);
+	return (SET_ERROR(rv));
```

Comment:
Normally SET_ERROR() appears where the error is first generated, where we see the EINVAL (or whatever) literal.  Is there a reason for doing it here?